File: app/src/main/java/com/example/recipeclipper/data/remote/RedditRecipeSource.kt
package com.example.recipeclipper.data.remote

import com.example.recipeclipper.data.Connectivity
import com.example.recipeclipper.data.model.ParseError
import com.example.recipeclipper.data.model.ParseResult
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.jsoup.Connection
import org.jsoup.HttpStatusException
import org.jsoup.Jsoup
import java.io.IOException
import java.net.SocketTimeoutException

/**
 * A Reddit post: one fetch of its public `.json` listing (the post and its comment tree
 * together), handed to [RedditRecipeParser]. A share link (`/r/<sub>/s/<code>`) is followed to
 * the post first, since only the post's own address has a listing.
 *
 * The endpoint is public and unauthenticated, so heavy use can meet HTTP 429: that is
 * [ParseError.Blocked], like any other refusal. Failures map to causes exactly as in
 * [BlogRecipeSource].
 */
class RedditRecipeSource(
    private val connectivity: Connectivity,
    private val timeoutMs: Int = 15_000,
    /** Where the listing is fetched from. A parameter only so a test can point it at a local
     *  responder. */
    private val base: String = RedditUrls.DEFAULT_BASE
) : RecipeSource {

    override suspend fun fetch(url: String): ParseResult = withContext(Dispatchers.IO) {
        try {
            val postUrl = if (RedditUrls.isShareLink(url)) {
                val response = connect(url).followRedirects(true).ignoreHttpErrors(true).execute()
                val landed = response.url().toString()
                if (RedditUrls.jsonUrl(landed, base) == null) {
                    val status = response.statusCode()
                    return@withContext ParseResult.Error(
                        if (status !in 200..299) ParseError.forHttpStatus(status) else ParseError.NoRecipeFound
                    )
                }
                landed
            } else {
                url
            }
            val jsonUrl = RedditUrls.jsonUrl(postUrl, base)
                ?: return@withContext ParseResult.Error(ParseError.NoRecipeFound)
            val body = connect(jsonUrl).ignoreContentType(true).execute().body()
            RedditRecipeParser.parse(body, url)
        } catch (e: HttpStatusException) {
            ParseResult.Error(ParseError.forHttpStatus(e.statusCode))
        } catch (e: CancellationException) {
            throw e // cancellation is never a failure to report
        } catch (e: IOException) {
            ParseResult.Error(
                when {
                    !connectivity.isOnline() -> ParseError.Offline
                    e is SocketTimeoutException -> ParseError.FetchFailed(e.message, timedOut = true)
                    else -> ParseError.FetchFailed(e.message)
                }
            )
        } catch (e: Exception) {
            ParseResult.Error(ParseError.FetchFailed(e.message))
        }
    }

    private fun connect(url: String): Connection = Jsoup.connect(url)
        .userAgent(USER_AGENT)
        .timeout(timeoutMs)
        .maxBodySize(MAX_BODY_BYTES)

    companion object {
        /** Reddit asks API clients to say who they are, in this shape. */
        const val USER_AGENT = "android:com.example.recipeclipper:1.0 (RecipeClipper)"

        /** A busy thread's listing runs to a few megabytes; Jsoup's default cap would cut it
         *  off mid-JSON. */
        const val MAX_BODY_BYTES = 16 * 1024 * 1024
    }
}

/**
 * The [RecipeSource] the repository sees: Reddit links go to [reddit], everything else to
 * [blog]. Chosen by host alone ([RedditUrls.isReddit]).
 */
class RoutingRecipeSource(
    private val blog: RecipeSource,
    private val reddit: RecipeSource
) : RecipeSource {
    override suspend fun fetch(url: String): ParseResult =
        if (RedditUrls.isReddit(url)) reddit.fetch(url) else blog.fetch(url)
}

File: ios/RecipeClipper/Data/Remote/RedditRecipeSource.swift
import Foundation

/// A Reddit post: one fetch of its public `.json` listing (the post and its comment tree
/// together), handed to `RedditRecipeParser`. A share link (`/r/<sub>/s/<code>`) is followed to
/// the post first, since only the post's own address has a listing. Android's
/// `RedditRecipeSource`.
///
/// The endpoint is public and unauthenticated, so heavy use can meet HTTP 429: that is
/// `.blocked`, like any other refusal. Failures map to causes exactly as in `BlogRecipeSource`.
final class RedditRecipeSource: RecipeSource {
    /// Reddit asks API clients to say who they are, in this shape.
    static let userAgent = "ios:com.example.recipeclipper:1.0 (RecipeClipper)"

    private let session: URLSession
    private let base: String

    /// `base` is where the listing is fetched from; a parameter only so a test can change it.
    init(session: URLSession = BlogRecipeSource.defaultSession, base: String = RedditUrls.defaultBase) {
        self.session = session
        self.base = base
    }

    func fetch(url: String) async -> ParseResult {
        var postUrl = url
        if RedditUrls.isShareLink(url) {
            switch await get(url) {
            case .failure(let error):
                return .error(error)
            case .success(let (_, landed, status)):
                guard RedditUrls.jsonUrl(landed, base: base) != nil else {
                    return .error((200..<300).contains(status) ? .noRecipeFound : ParseError.forHttpStatus(status))
                }
                postUrl = landed
            }
        }
        guard let jsonUrl = RedditUrls.jsonUrl(postUrl, base: base) else { return .error(.noRecipeFound) }
        switch await get(jsonUrl) {
        case .failure(let error):
            return .error(error)
        case .success(let (data, _, status)):
            guard (200..<300).contains(status) else { return .error(ParseError.forHttpStatus(status)) }
            return RedditRecipeParser.parse(String(decoding: data, as: UTF8.self), sourceUrl: url)
        }
    }

    /// One GET, following redirects: the body, the address it finally came from, and the
    /// status. A transport failure is already a cause.
    private func get(_ address: String) async -> Result<(Data, String, Int), ParseError> {
        guard let target = URL(string: address),
              let scheme = target.scheme?.lowercased(), scheme == "http" || scheme == "https"
        else { return .failure(.fetchFailed(URLError(.badURL).localizedDescription)) }
        var request = URLRequest(url: target, timeoutInterval: BlogRecipeSource.timeout)
        request.httpMethod = "GET"
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        do {
            let (data, response) = try await session.data(for: request)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 200
            return .success((data, response.url?.absoluteString ?? address, status))
        } catch {
            // As BlogRecipeSource: a cancelled fetch reports nothing; the repository discards it.
            if error is CancellationError || (error as? URLError)?.code == .cancelled {
                return .failure(.fetchFailed(nil))
            }
            return .failure(BlogRecipeSource.cause(of: error))
        }
    }
}

/// The `RecipeSource` the repository sees: Reddit links go to `reddit`, everything else to
/// `blog`. Chosen by host alone (`RedditUrls.isReddit`).
struct RoutingRecipeSource: RecipeSource {
    let blog: RecipeSource
    let reddit: RecipeSource

    func fetch(url: String) async -> ParseResult {
        RedditUrls.isReddit(url) ? await reddit.fetch(url: url) : await blog.fetch(url: url)
    }
}
